model/sharing/sharing.go

```diff
@@ -1055,206 +1057,635 @@ func CheckSharings(inst *instance.Instance) ([]map[string]interface{}, error) {
+				// assume the sharing synchronization is still in progress and
+				// that would explain why it's missing on the other instance.
+				// In this case, we do nothing.
+				continue
 			}
 
-			if !s.ReadOnly() && s.Members[0].Instance == "" {
+			if isFileTooBigForInstance(inst, memberDoc) {
 				checks = append(checks, map[string]interface{}{
-					"id":     s.SID,
-					"type":   "missing_instance_for_member",
-					"member": 0,
+					"id":       s.SID,
+					"type":     "disk_quota_exceeded",
+					"instance": inst.Domain,
+					"file":     memberDoc,
 				})
+				continue
 			}
+
+			checks = append(checks, map[string]interface{}{
+				"id":      s.SID,
+				"type":    "missing_matching_doc_for_owner",
+				"member":  m.Domain,
+				"missing": memberDoc,
+			})
 		}
+	}
 
-		return nil
-	})
-	return checks, err
+	// The only docs left in the map do not exist on the member's instance
+	for _, ownerDoc := range ownerDocsById {
+		if wasUpdatedRecently(ownerDoc) {
+			// If the document was created less than 5 minutes ago, we'll
+			// assume the sharing synchronization is still in progress and
+			// that would explain why it's missing on the other instance.
+			// In this case, we do nothing.
+			continue
+		}
+
+		if isFileTooBigForInstance(m, ownerDoc) {
+			checks = append(checks, map[string]interface{}{
+				"id":       s.SID,
+				"type":     "disk_quota_exceeded",
+				"instance": m.Domain,
+				"file":     ownerDoc,
+			})
+			break
+		}
+
+		checks = append(checks, map[string]interface{}{
+			"id":         s.SID,
+			"type":       "missing_matching_doc_for_member",
+			"member":     m.Domain,
+			"missing":    ownerDoc,
+			"ownerDocID": ownerDoc.ID(),
+		})
+	}
+
+	return checks
+}
+
+// isFileTooBigForInstance returns true if the given doc represents a file and
+// its size is greater than the available space on the given instance.
+// If said instance does not have any defined quota, it returns false.
+func isFileTooBigForInstance(inst *instance.Instance, doc couchdb.JSONDoc) bool {
+	if docType, ok := doc.M["type"].(string); !ok || docType == "" || docType == consts.DirType {
+		return false
+	}
+
+	var file *vfs.FileDoc
+
+	fileJSON, err := json.Marshal(doc)
+	if err != nil {
+		return false
+	}
+
+	if err := json.Unmarshal(fileJSON, &file); err != nil {
+		return false
+	}
+
+	_, _, _, err = vfs.CheckAvailableDiskSpace(inst.VFS(), file)
+	return err == vfs.ErrFileTooBig
+}
+
+// wasUpdatedRecently returns true if the given document's latest update, given
+// by its `cozyMetadata.updatedAt` attribute, happened less than 5 minutes ago.
+// If the attribute is missing or does not represent a valid date, we consider
+// the latest update happened before that.
+func wasUpdatedRecently(doc couchdb.JSONDoc) bool {
+	cozyMetadata, ok := doc.M["cozyMetadata"].(map[string]interface{})
+	if !ok || cozyMetadata == nil {
+		return false
+	}
+	if updatedAt, ok := cozyMetadata["updatedAt"].(time.Time); ok {
```

Comment:
I'm a bit surprised to have a `time.Time` here. I would have expected it to be a string, as the JSON was unmarshaled to a `map[string]interface{}`, not to a `struct` that would say that its type is a `time.Time`.

Reply:
I was surprised as well and the first implementation was parsing a string to get a time value but I got type errors.
It makes sense when you follow the trail though:
1. https://github.com/cozy/cozy-stack/pull/3597/files#diff-d96f65ea3d1c149ba02f26b9d45c1a92b040deda89c41acb212a1947b776ac21R252-R256
2. https://github.com/cozy/cozy-stack/blob/master/model/sharing/files.go#L1272 & https://github.com/cozy/cozy-stack/blob/master/model/sharing/files.go#L1311
3. https://github.com/cozy/cozy-stack/blob/e15da402ff79315503d1f659a3dcb83c1a3f0f4f/model/vfs/cozy_metadata.go#L118